protocol-correctness/lib/language/map-utils.k

```diff
@@ -0,0 +1,165 @@
+require "protocol-correctness/lib/language/maybe.k"
+require "protocol-correctness/lib/language/reflection.k"
+require "protocol-correctness/lib/language/type-reflection.k"
+
+module MAP-UTILS-SYNTAX
+  imports MAP
+  imports TYPE-REFLECTION-SYNTAX
+
+  syntax KItem ::= concretizeMapKey(KItem, Map)
+
+  syntax KItem ::= nullableMapLookup(key:KItem, Map, ReflectionType)
+  syntax KItem ::= nullableMapSet(key:KItem, value:KItem, Map, ReflectionType)
+endmodule
+
+module MAP-UTILS
+  imports MAP-REFLECTION
+  imports MAP-UTILS-BASIC
+  imports MAP-UTILS-LANGUAGE
+endmodule
+
+module MAP-UTILS-SYNTAX-PRIVATE
+  imports MAP
+
+  syntax KItem ::= mapLookup(KItem, Map)
+  syntax KItem ::= mapDelete(key:KItem, Map)
+  syntax KItem ::= mapSet(key:KItem, value:KItem, Map)
+endmodule
+
+module MAP-UTILS-BASIC
+  imports BOOL
+  imports K-EQUAL-SYNTAX
+  imports MAP
+  imports MAP-UTILS-SYNTAX
+  imports MAP-UTILS-SYNTAX-PRIVATE
+  imports MAYBE
+  imports TYPE-REFLECTION-SYNTAX
+
+  rule concretizeMapKey(Key:KItem, M:Map) => splitMap(Key, M, ?_Value, ?_Remainder)
+    requires Key in_keys(M)
+    [label(concretizeMapKey)]
+  rule concretizeMapKey(Key:KItem, M:Map) => .K
+    requires notBool Key in_keys(M)
+    [label(concretizeKeyNotInMap)]
+
+
+  syntax KItem ::= "splitMapStuck" | "endSplitMap"
+  rule endSplitMap => .K
+
+
+  syntax KItem ::= splitMap(key:KItem, toSplit:Map, value:KItem, remainder:Map)
+      [function, functional]
+  rule splitMap(Key:KItem, M:Map, _Value:KItem, _Remainder:Map) => splitMapStuck
+    requires notBool Key in_keys(M)
+  rule splitMap(Key:KItem, (Key |-> _:KItem) M:Map, _Value:KItem, _Remainder:Map)
+      => endSplitMap
+    ensures notBool Key in_keys(M)
+    [simplification(30)]
+  rule splitMap(Key:KItem, (K1:KItem |-> _SomeValue:KItem) M:Map, Value:KItem, Remainder:Map)
+      => splitMap(Key, M, Value, Remainder)
+    requires notBool (Key ==K K1) andBool Key in_keys(M)
+    ensures notBool K1 in_keys(M)
+    [simplification(30)]
+  rule splitMap(Key:KItem, M:Map, Value:KItem, Remainder:Map)
+      => endSplitMap
+    requires Key in_keys(M)
+    ensures M ==K (Key |-> Value Remainder)
+    [simplification(50)]
+
+  // Do the map lookups in a way that makes the Haskell backend happy.
+
+  rule mapLookup(K:KItem, M:Map)
+      =>  concretizeMapKey(K, M) ~> #mapLookup(K, M)
+    requires K in_keys(M)
+    [label(mapLookupInKeys)]
+  rule mapLookup(K:KItem, M:Map)
+      =>  nothing
+    requires notBool K in_keys(M)
+    [label(mapLookupNotInKeys)]
+
+  syntax KItem ::= #mapLookup(KItem, Map)
+  rule #mapLookup(K:KItem, ((K |-> V:KItem) _:Map) #as M:Map) => just(V)
+    ensures K in_keys(M)
+  rule #mapLookup(K:KItem, M:Map) => nothing
+    ensures notBool (K in_keys(M))
+    [owise]
+
+  rule mapDelete(K:KItem, M:Map) => #mapDelete(K, M)
```

Comment:
Do you need to concretize the key here? The final result will be ?M:Map and ensures notBool K in_keys(?M)

Reply:
See the comment below.